sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -1080,136 +1121,484 @@ private static Progress scaleProgress(
+                newWindowStopIndex,
+                stopWindowIndex,
+                null,
+                watermarkAndState);
+      } else {
+        // Compute the element split with the scaled fraction.
+        SplitResult<?> elementSplit =
+            currentTracker.trySplit(scaledFractionOfRemainder / elementProgress.getWorkRemaining());
+        newWindowStopIndex = currentWindowIndex + 1;
+        if (elementSplit != null) {
+          windowedSplitResult =
+              computeWindowSplitResult(
+                  currentElement,
+                  currentRestriction,
+                  currentWindow,
+                  windows,
+                  currentWatermarkEstimatorState,
+                  currentWindowIndex,
+                  newWindowStopIndex,
+                  stopWindowIndex,
+                  elementSplit,
+                  watermarkAndState);
+        } else {
+          windowedSplitResult =
+              computeWindowSplitResult(
+                  currentElement,
+                  currentRestriction,
+                  currentWindow,
+                  windows,
+                  currentWatermarkEstimatorState,
+                  newWindowStopIndex,
+                  newWindowStopIndex,
+                  stopWindowIndex,
+                  null,
+                  watermarkAndState);
+        }
+      }
+    } else {
+      // We are on the last window then compute the element split with given fraction.
+      newWindowStopIndex = stopWindowIndex;
+      SplitResult<?> splitResult = currentTracker.trySplit(fractionOfRemainder);
+      if (splitResult == null) {
+        return null;
+      }
+      windowedSplitResult =
+          computeWindowSplitResult(
+              currentElement,
+              currentRestriction,
+              currentWindow,
+              windows,
+              currentWatermarkEstimatorState,
+              currentWindowIndex,
+              stopWindowIndex,
+              stopWindowIndex,
+              splitResult,
+              watermarkAndState);
+    }
+    return KV.of(windowedSplitResult, newWindowStopIndex);
+  }
+
   private HandlesSplits.SplitResult trySplitForElementAndRestriction(
       double fractionOfRemainder, Duration resumeDelay) {
     KV<Instant, WatermarkEstimatorStateT> watermarkAndState;
-    WindowedSplitResult windowedSplitResult;
+    WindowedSplitResult windowedSplitResult = null;
     synchronized (splitLock) {
       // There is nothing to split if we are between element and restriction processing calls.
       if (currentTracker == null) {
         return null;
       }
-
       // Make sure to get the output watermark before we split to ensure that the lower bound
       // applies to the residual.
       watermarkAndState = currentWatermarkEstimator.getWatermarkAndState();
-      SplitResult<RestrictionT> splitResult = currentTracker.trySplit(fractionOfRemainder);
+      KV<WindowedSplitResult, Integer> splitResult =
```

Comment:
You might want to make use of an `@AutoValue` for the return type to make it clear the updated stop index is being returned.

Reply:
Let's leave it as it is and I'll update it together with clean-up?